text-transform test failures due to Lithuanian rules and out-of-flow characters

darinadler opened the pull request:
#### 6e67536ebe5b6515f8086dfb4111378655de0bbd
<pre>
text-transform test failures due to Lithuanian rules and out-of-flow characters
<a href="https://bugs.webkit.org/show_bug.cgi?id=268705">https://bugs.webkit.org/show_bug.cgi?id=268705</a>
<a href="https://rdar.apple.com/122251072">rdar://122251072</a>

Reviewed by Tim Nguyen.

Changed so Lithuanian won&apos;t use the fast path, and so we won&apos;t consider out of line
characters when doing title case.

* LayoutTests/TestExpectations: Expect 3 more tests to pass. Also took off incorrect bug
numbers. Many tests were marked with a closed bug, which is not what bug numbers in this
file are supposed to mean.

* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::needsTurkishCasingRules): Tweaked style.
(WTF::needsGreekUppercasingRules): Ditto.
(WTF::needsLithuanianCasingRules): Added.
(WTF::StringImpl::convertToLowercaseWithLocale): Don&apos;t use fast path for Lithuanian.
(WTF::StringImpl::convertToUppercaseWithLocale): Ditto.

* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::previousCharacter const): Skip renderers that are not in flow.

Canonical link: <a href="https://commits.webkit.org/274073@main">https://commits.webkit.org/274073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a1f2a34fcbefc4b48b50f2b2dcdb9f5fe78ec86

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16669 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40015 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40314 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33601 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/39101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19317 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13893 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31957 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38341 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14020 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33058 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12257 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41573 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/31492 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34169 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34184 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38078 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/37474 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12784 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10352 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36253 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14196 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/44387 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8490 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13164 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9073 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13507 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1477,9 +1477,10 @@ static inline bool isInlineFlowOrEmptyText(const RenderObject& renderer)
-    // find previous text renderer if one exists
     const RenderObject* previousText = this;
     while ((previousText = previousText->previousInPreOrder())) {
+        if (!previousText->isInFlow())
+            continue;
```
nit: this is slightly confusing as it implies there can be out-of-flow text (a box can be out-of-flow but that does not make the text inside it out-of-flow).
`while ((previousText = previousText->previousInPreOrder())) {`
should probably read
`while ((previousSibling = previousSibling->previousInPreOrder())) {`
again, just a nit.

darinadler replied:
Yes, not something I introduced, but I agree the style in the function is confusing, with a variable named `previousText` that is not a text renderer. I hope you’re OK that I didn’t fix that. I don’t think the `isInFlow` check is particularly more confusing than the other things happening there, like checking if the `previousText` is a text renderer or not, so I will leave it as is for now.

alanbaradlay replied:
Absolutely. I am going to clean this up once your PR lands. I also noticed that the static helper we call here was incorrectly named `isInlineFlowOrEmptyText` -> `isInlineBoxOrEmptyText`

darinadler replied:
I also wonder whether this logic is right for more unusual cases. The new code I added might be wrong if the text being capitalized is itself inside an out of flow renderer.

alanbaradlay replied:
I think this is more correct than not. Content inside an out-of-flow box is usually means it is "detached" from the rest of the flow (and in most cases it ends up somewhere else on the screen) and same applies to floating positioned content.

darinadler committed:
text-transform test failures due to Lithuanian rules and out-of-flow characters
https://bugs.webkit.org/show_bug.cgi?id=268705
rdar://122251072

Reviewed by Tim Nguyen.

Changed so Lithuanian won't use the fast path, and so we won't consider out of line
characters when doing title case.

* LayoutTests/TestExpectations: Expect 3 more tests to pass. Also took off incorrect bug
numbers. Many tests were marked with a closed bug, which is not what bug numbers in this
file are supposed to mean.

* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::needsTurkishCasingRules): Tweaked style.
(WTF::needsGreekUppercasingRules): Ditto.
(WTF::needsLithuanianCasingRules): Added.
(WTF::StringImpl::convertToLowercaseWithLocale): Don't use fast path for Lithuanian.
(WTF::StringImpl::convertToUppercaseWithLocale): Ditto.

* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::previousCharacter const): Skip renderers that are not in flow.

Canonical link: https://commits.webkit.org/274073@main
